common/optimiser/OptimiserConstraint.java

```diff
@@ -31,24 +31,35 @@ public class OptimiserConstraint {
     private final String name;
     private final Map<OptimiserVariable<?>, Double> coefficients;
     private MPConstraint mpConstraint;
+    boolean isInitialised;
 
     public OptimiserConstraint(double lowerBound, double upperBound, String name) {
         this.lowerBound = lowerBound;
         this.upperBound = upperBound;
         this.coefficients = new HashMap<>();
         this.name = name;
+        this.isInitialised = false;
     }
 
     public void setCoefficient(OptimiserVariable<?> variable, double coeff) {
+        if (isInitialised) {
+            if (coefficients.containsKey(variable)) {
+                isInitialised = coefficients.get(variable) == coeff;
+            } else {
+                isInitialised = false;
+            }
+        }
         coefficients.put(variable, coeff);
     }
 
     synchronized void initialise(MPSolver solver) {
-        this.mpConstraint = solver.makeConstraint(lowerBound, upperBound, name);
+        if (this.mpConstraint == null) this.mpConstraint = solver.makeConstraint(lowerBound, upperBound, name);
         coefficients.forEach((var, coeff) -> mpConstraint.setCoefficient(var.mpVariable(), coeff));
+        isInitialised = true;
```

Comment:
I think you're using the `isInitialised` to do too many things here (and this is compounded by the naming). If we rename `isInitialised` to `coefficientsChanged` or `coefficientsApplied` or `coefficientsSet`, we realise that the `setCoefficient` method can be simplified massively to:
```

    public void setCoefficient(OptimiserVariable<?> variable, double coeff) {
        if (mpConstraint != null) coefficientsApplied = coefficients.get(variable) == coeff;
        coefficients.put(variable, coeff);
    }
```

The new naming also indicates that we should set it to `false` when we `release()` if that doesn't affect the logical integrity.

(note that the `.get()` will return a null if it doesn't exist which will not match the value you pass in anyway)

This way, we use one signal (`mpConstraint` being null or not) to indicate whether we've initialised with the previous set of constraints, and then the `coefficientsApplied` to record whether active constraint contains any outdated coefficients